### PR TITLE
Tz/kfp code packaging

### PR DESCRIPTION
**Pull request by talebzeghmi**
* Work to make Metaflow work on KFP with Docker image already containing the code.
* parameterized `datastore_root` KFP pipeline variable.

### CODE REVIEW

File: metaflow/plugins/kfp/kfp.py [hunk 1]
```diff
@@ -78,12 +85,15 @@ def create_kfp_pipeline_yaml(self, pipeline_file_path=DEFAULT_KFP_YAML_OUTPUT_PA
...
-    def _command(code_package_url, environment, step_name, step_cli):
+    def _command(self, code_package_url, environment, step_name, step_cli):
         """
         Analogous to batch.py
         """
-        commands = environment.get_package_commands(code_package_url)
+        commands = (
+            environment.get_package_commands(code_package_url)
+            if self.s3_code_package
+            else ["cd " + str(Path(inspect.getabsfile(self.flow.__class__)).parent)]
```
Comment (alexlatchford) [inline metaflow/plugins/kfp/kfp.py:95]: So guessing the `s3_code_package` pieces are just to maintain the compatibility with the existing Metaflow patterns but we just intend to use the `base_image` which will contain the Code Package already?

File: metaflow/plugins/kfp/kfp.py [hunk 2]
```diff
@@ -358,33 +368,34 @@ def build_kfp_dag(node: DAGNode, context, index=None):
...
     from typing import NamedTuple
 
     print("----")
     print("context")
-    print(type(contexts))
-    print(contexts)
+    print(context)
     print("----")
-    context_dict = json.loads(contexts)
+    context_dict = json.loads("{}" if context == "" else context)
```
Comment (alexlatchford) [inline metaflow/plugins/kfp/kfp.py:393]: Maybe I missed this in earlier PRs but the `context` here is injected at runtime by KFP?

Just asking as it looks like we also set this at compile time (I think) during `build_kfp_dag` which I'm guessing cannot actually resolve the `context_dict.get("task_id", "")`.

Or is it actually executed both at compile and run time?

**Review by alexlatchford — APPROVED** [review]

File: metaflow/plugins/kfp/kfp.py (hunk 1 above)
Comment (talebzeghmi) [reply metaflow/plugins/kfp/kfp.py:95]: yes, the s3_code_package allows one to run `python hello.py --datastore=s3 kfp run` from their laptop onto KFP without needing to package hello.py into a Docker container!

example:
```bash
```

File: metaflow/plugins/kfp/kfp.py (hunk 2 above)
Comment (talebzeghmi) [reply metaflow/plugins/kfp/kfp.py:393]: True, context is being passed along to get the parent taskId at runtime, 

True this [code](https://github.com/zillow/metaflow/blob/8c2913661d7eb5bb5417500f9da4bca83bd3dcd4/metaflow/plugins/kfp/kfp.py#L161-L207) will deprecate soon, especially the [current_task_id](https://github.com/zillow/metaflow/blob/8c2913661d7eb5bb5417500f9da4bca83bd3dcd4/metaflow/plugins/kfp/kfp.py#L161-L161).  

TaskIds will be defined at runtime by KFP/Argo.  This design of knowing the taskIds at compile time doesn't work with foreach, where the number of splits is only known at runtime.

**Review by cloudw — APPROVED** [review]

File: metaflow/plugins/kfp/kfp.py [hunk 3]
```diff
@@ -278,8 +288,8 @@ def _step_cli(self, node, input_paths, user_code_retries, task_id):
             "--quiet",
             "--metadata=%s" % self.metadata.TYPE,
             "--environment=%s" % self.environment.TYPE,
-            "--datastore=%s" % self.datastore.TYPE,
-            "--datastore-root=%s" % self.datastore.datastore_root,
+            "--datastore=s3",
```
Comment (hsezhiyan) [inline metaflow/plugins/kfp/kfp.py:291]: Does this mean the datastore is defaulted to s3? If the user wants to change the datastore, where could they do it?

File: metaflow/plugins/kfp/kfp_cli.py [hunk 4]
```diff
@@ -46,38 +46,59 @@ def kubeflow_pipelines(obj):
...
 )
 @click.option(
     "--pipeline-path",
     "pipeline_path",
     default=DEFAULT_KFP_YAML_OUTPUT_PATH,
-    help="the output path (or filename) of the generated KFP pipeline yaml file",
+    help="The output path of the generated KFP pipeline yaml file",
+)
+@click.option(
+    "--s3-code-package/--no-s3-code-package",
```
Comment (hsezhiyan) [inline metaflow/plugins/kfp/kfp_cli.py:84]: Why do you also `--no-s3-code-package` here? There should be just one flag, `s3-code-package`, which indicates whether or not the code package is saved in s3.

File: metaflow/plugins/kfp/kfp.py (hunk 3 above)
Comment (talebzeghmi) [reply metaflow/plugins/kfp/kfp.py:291]: KFP for now supports only supports s3 datastore.  This is how Metaflow state can be accessed between KFP step containers.

File: metaflow/plugins/kfp/kfp_cli.py (hunk 4 above)
Comment (talebzeghmi) [reply metaflow/plugins/kfp/kfp_cli.py:84]: This is the click way to enable or disable a flag type argument. 
see https://click.palletsprojects.com/en/7.x/options/#boolean-flags

```python
import sys

@click.command()
@click.option('--shout/--no-shout', default=False)
def info(shout):
    rv = sys.platform
    if shout:
        rv = rv.upper() + '!!!!111'
    click.echo(rv)
```

**Pull request merged**